solanaj/src/main/java/org/p2p/solanaj/ws/SubscriptionWebSocketClient.kt

```diff
@@ -0,0 +1,143 @@
+package org.p2p.solanaj.ws
+
+import android.util.Log
+import com.squareup.moshi.Moshi
+import com.squareup.moshi.Types
+import org.java_websocket.WebSocket
+import org.java_websocket.client.WebSocketClient
+import org.java_websocket.exceptions.WebsocketNotConnectedException
+import org.java_websocket.framing.Framedata
+import org.java_websocket.handshake.ServerHandshake
+import org.p2p.solanaj.model.types.RpcNotificationResult
+import org.p2p.solanaj.model.types.RpcRequest
+import org.p2p.solanaj.model.types.RpcResponse
+import java.net.URI
+import java.net.URISyntaxException
+
+// TODO: Refactor this class, move to kotlin, handle 1006 error, maybe ping/pong issue
+class SubscriptionWebSocketClient(serverURI: URI?) : WebSocketClient(serverURI) {
+    private inner class SubscriptionParams constructor(
+        var request: RpcRequest,
+        var listener: NotificationEventListener
+    )
+
+    private val subscriptions = mutableMapOf<String, SubscriptionParams>()
+    private val subscriptionIds = mutableMapOf<String, Long?>()
+    private val subscriptionListeners = mutableMapOf<Long, NotificationEventListener>()
+    fun ping() {
+        try {
+            if (instance?.isOpen == true) instance?.sendPing()
+        } catch (error: WebsocketNotConnectedException) {
+            Log.e("SOCKET", "Error on ping socket", error)
+        }
+    }
+
+    fun accountSubscribe(key: String, listener: NotificationEventListener) {
+        val rpcRequest = RpcRequest("accountSubscribe", mutableListOf(key))
+        subscriptions[rpcRequest.id] = SubscriptionParams(rpcRequest, listener)
+        subscriptionIds[rpcRequest.id] = null
+        updateSubscriptions()
+    }
+
+    fun signatureSubscribe(signature: String, listener: NotificationEventListener) {
+        val rpcRequest = RpcRequest("signatureSubscribe", mutableListOf(signature))
+        subscriptions[rpcRequest.id] = SubscriptionParams(rpcRequest, listener)
+        subscriptionIds[rpcRequest.id] = null
+        updateSubscriptions()
+    }
+
+    override fun onWebsocketPong(socket: WebSocket, framedata: Framedata) {
+        super.onWebsocketPong(socket, framedata)
+        socketStateListener?.onWebSocketPong()
+    }
+
+    override fun onOpen(handshakedata: ServerHandshake) {
+        socketStateListener?.onConnected()
+        updateSubscriptions()
+    }
+
+    override fun onMessage(message: String) {
+        Log.d("SOCKET", "New message received: $message")
+        val resultAdapter = Moshi.Builder().build()
```

Comment:
maybe create moshi as a field or extract adapter creation? it would allow us to go from wrapping this line